### PR TITLE
Jnj integration

### PR DESCRIPTION
**Pull request by devhas**

### CODE REVIEW

File: app/controllers/webhooks_controller.rb [hunk 1]
```diff
@@ -1,29 +1,32 @@
...
-  # TEST: `FactoryBot.create :work_instruction_session, company: Company.first`
-  # - Company.first.update_attributes webhook: 'http://localhost:3031/webhooks/receive'
-  # - try wget << ipconfig getifaddr en0 >>:3031 from inside container
-  def receive
-    # Step 1: save the incoming payload just in case further processing fails
-    @webhook = Webhook.create(payload: payload, origin: request.remote_ip)
+    origin = request.headers["origin"]
+    if origin.include?("localhost")
+      url = "https://corpisnad.jnj.com/rest/JU001_LMS_CSOD_CourseCompletion_RESTWS/wsInbound/ws?category=createExpressClass"
+    elsif orign.includes("cms.scopear.com")
```
Comment (devhas) [inline app/controllers/webhooks_controller.rb:17]: Please let me know the QA and production origin so that I can add the check accordingly.

File: app/services/api_service.rb [hunk 2]
```diff
@@ -1,125 +1,30 @@
...
-  API_WIS_URL               = API_URL + 'v2/ar_contents/:ar_content_id/work_instruction_sessions'
-
-  #The url that you'll use to execute a GraphQL query against
-  API_GQL_URL               = API_URL + 'v3/graphql'
-
-  # The path that you'll use to authenticate
-  API_AUTH_URL              = API_URL + 'v2/users/sign_in.json'
+  # jnj server creds
+  JNJ_Username = "JU001_LMS_WorkLink"
+  JNJ_Password = "JU001_LMS_WorkLink"
```
Comment (devhas) [inline app/services/api_service.rb:4]: Please provide the password for production authentication. 
https://corpisnad.jnj.com/rest/JU001_LMS_CSOD_CourseCompletion_RESTWS/wsInbound/ws?category=createExpressClass
Username : JU001_LMS_WorkLink
Password : JU001_LMS_WorkLink


JnJ QA End point:
https://corpisnaq.jnj.com/rest/JU001_LMS_CSOD_CourseCompletion_RESTWS/wsInbound/ws?category=createExpressClass

Username : JU001_LMS_Worklink
Password : D76arsa@23gyHXb#

QA and dev creds are given but production creds are not given. 
Ref: https://scopearcloud.atlassian.net/browse/CMS-3981?focusedCommentId=45325

File: app/services/api_service.rb [hunk 3]
```diff
@@ -1,125 +1,30 @@
...
-            variables: variables
-          }.to_json,
-          headers: HEADERS.merge("PrivateAccessCode" => "Token token=#{TWO_FACTOR_TOKEN}")
-        )
-      end
+    def send_request_jnj(email:, url:)
+      try_twice(
+        uri: URI.parse(url),
+        form_data: {
+          email: email
```
Comment (devhas) [inline app/services/api_service.rb:16]: Do we currently only need to provide email in the body and should we just hard code the status value as passed/ completed or do we need any other attribute also.

File: app/services/api_service.rb (hunk 2 above)
Comment (devhas) [reply app/services/api_service.rb:4]: https://share.getcloudapp.com/RBul94Lb

Comment (amazing-jay) [reply app/services/api_service.rb:4]: I've asked for them
https://share.getcloudapp.com/E0uAQxkn

File: README.md [hunk 4]
```diff
@@ -5,121 +5,9 @@ Intended to be replace for use with JNJ
 
   This Scope AR API-Demo-Ruby is a simple Ruby-On-Rails client application that demonstrates a complete, end-to-end integration with Scope AR APIs.
```
Comment (amazing-jay) [inline README.md:7]: this isn't accurate

File: README.md [hunk 5]
```diff
@@ -5,121 +5,9 @@ Intended to be replace for use with JNJ
...
-    4. Run `rails s -p 3031` (NOTE: any non-standard port suffices to prevent conflicts with local API server)
-    5. Open a browser to http://localhost:3031
-
-  Remote API Server:
-
-    1. Clone this repo
-    2. Configure client & server applications (SEE "Client Configuration" & "Server Configuration" Sections)
-    3. Run `rails db:setup`
-    4. Run `rails s`
-    5. Open a browser to http://your.custom.ip.or.domain
```
Comment (amazing-jay) [inline README.md:50]: why did you delete setup?

File: README.md [hunk 6]
```diff
@@ -5,121 +5,9 @@ Intended to be replace for use with JNJ
...
-
 # Important Files
 
   * [app/controllers/webhooks_controller.rb](/tree/develop/app/controllers/webhooks_controller.rb)
-  * [app/jobs/get_resource_info_job.rb](/tree/develop/app/jobs/get_resource_info_job.rb)
-  * [app/models/auth_token.rb](/tree/develop/app/models/auth_token.rb)
-  * [app/services/api_service.rb](/tree/develop/app/services/api_service.rb)
-  * [config/credentials.yml.enc](/tree/develop/config/credentials.yml.enc)
-
-# System dependencies
```
Comment (amazing-jay) [inline README.md:24]: Why did you delete all the dependency and setup info? seems like its all still relevant until #Retrieving the ARContentId.

Seems like you should restore and then update to make sure still accurate

File: app/controllers/webhooks_controller.rb [hunk 7]
```diff
@@ -1,29 +1,32 @@
...
-  def recent
-    @limit = 10
-    @webhooks = Webhook.limit(@limit).order(id: :desc)
-  end
+    @webhook = WebhookTopic.find_by(webhook_topic: webhook_topic, user_id: user_id)
+    if @webhook.present?
+      return head :ok
+    else
+      @webhook = WebhookTopic.create(webhook_topic: webhook_topic, user_id: user_id)
+    end
```
Comment (amazing-jay) [inline app/controllers/webhooks_controller.rb:12]: we want to respond back with head :ok regardless if found or not found; as fast as possible; and kick off the request to the client as a background job; something like:
```
@webhook = WebhookTopic.find_or_create... 
if @webhook.valid? & @webhook.persisted?
  ClientRequestWorker.perform(@webhook) 
  return head 🆗 
else
  return head :500
end
```

File: README.md [hunk 8]
```diff
@@ -129,20 +17,3 @@ Intended to be replace for use with JNJ
   Production Mode:
   * mysql
   * sidekiq
-
-# Deployment instructions
```
Comment (amazing-jay) [inline README.md:133]: update with instruction specific for heroku

File: app/controllers/webhooks_controller.rb (hunk 7 above)
Comment (amazing-jay) [reply app/controllers/webhooks_controller.rb:12]: read this: https://scopearcloud.atlassian.net/wiki/spaces/API/pages/1270448421/Using+WebHooks#Frequency

File: app/controllers/webhooks_controller.rb (hunk 1 above)
Comment (amazing-jay) [reply app/controllers/webhooks_controller.rb:17]: https://scopearcloud.atlassian.net/browse/CMS-3980

note -> we often link relevant tickets in Jira; try looking in Jira first (the above ticket was linked to your ticket)

Comment (amazing-jay) [reply app/controllers/webhooks_controller.rb:17]: also, we'll have to find time after I land in greece to show you how to use the mobile app for testing

File: app/services/api_service.rb (hunk 3 above)
Comment (amazing-jay) [reply app/services/api_service.rb:16]: same feedback as earlier; read all information in Jira tickets, comments, and linked documents first.

[This document is linked](https://drive.google.com/file/d/1InEltXaOxWoqDtL8krAySqKRomqB92n4/view) to your ticket in a recent comment. 

[Page 4 shows a sample request payload](https://share.getcloudapp.com/2NuqZg24)

Note that their endpoint supports a list of updates, but we don't need to batch. We can send each update as a different request.

File: config/environments/development.rb [hunk 9]
```diff
@@ -11,6 +11,8 @@
   # Do not eager load code on boot.
   config.eager_load = false
 
+  config.hosts << /[a-z0-9]+\.ngrok\.io/
```
Comment (amazing-jay) [inline config/environments/development.rb:14]: Problem: not all developers will want to use ngrok.
Solution: move this into a configuration file. 
Note: I believe this is a rails 6 project, which uses credentials rather than dotenv. See the instructions that you deleted in the README to learn how to update the credentials file.

File: README.md [hunk 10]
```diff
@@ -5,121 +5,9 @@ Intended to be replace for use with JNJ
...
-  Run `rails credentials:edit` to configure this application.
-
-  The encrypted crendentials file includes two sets of default values which can be used 'as is' to run the application \
-  against either: a) an instance of the API server running on localhost, b) Scope AR's production cloud servers.
-
-  Comment out the "other" set of values (as indicated in the file to use either of the default configurations.
-  Contact api.support@scopear.com to use a custom server (or custom Organization).
-
-  **Note #1: `config/master.key` is atypically checked into source control so that you can easily decrypt and edit the credentials file.**
-  **Note #2: the local development option requires additional steps to be performed on the API server (see "Server Configuration").**
```
Comment (amazing-jay) [inline README.md:63]: This part is relevant for a comment I left on this PR later. Note that because this is no longer a public repo, you'll want to remove the config/master.key from source control. Also, you'll want to move the development credentials file to a sample variant which each developer is suppose to copy for local env.

see: https://blog.saeloun.com/2019/10/10/rails-6-adds-support-for-multi-environment-credentials.html

File: app/controllers/webhooks_controller.rb [hunk 11]
```diff
@@ -1,29 +1,31 @@
...
-  # - Company.first.update_attributes webhook: 'http://localhost:3031/webhooks/receive'
-  # - try wget << ipconfig getifaddr en0 >>:3031 from inside container
   def receive
-    # Step 1: save the incoming payload just in case further processing fails
-    @webhook = Webhook.create(payload: payload, origin: request.remote_ip)
+    webhook_topic = request.headers["X-Scope-Topic"]
+    if webhook_topic_params["node"]
+      user_id = webhook_topic_params["node"]["user"]["username"] if webhook_topic_params["node"]["user"]
+      email = webhook_topic_params["node"]["user"]["email"] if webhook_topic_params["node"]["user"]
+      object_id = webhook_topic_params["node"]["id"]
```
Comment (amazing-jay) [inline app/controllers/webhooks_controller.rb:7]: note to self: need to finialize spec -> where will the JnJ object id be stored? will it be in the Scenario.description field or the Scenario.externalData.objectId field (and what will the final attribute name be)?

File: app/controllers/webhooks_controller.rb [hunk 12]
```diff
@@ -1,29 +1,31 @@
...
+      state = webhook_topic_params["node"]["state"]
```
Comment (amazing-jay) [inline app/controllers/webhooks_controller.rb:8]: get rid of this check

Comment (amazing-jay) [reply app/controllers/webhooks_controller.rb:8]: also ask for org_id = webhook_topic_params["node"]["organization"]["id"]

File: app/controllers/webhooks_controller.rb [hunk 13]
```diff
@@ -1,29 +1,31 @@
...
-    @webhooks = Webhook.limit(@limit).order(id: :desc)
-  end
-
-  # POST /webhooks/receive or /webhooks/receive.json
-  # TEST: `FactoryBot.create :work_instruction_session, company: Company.first`
-  # - Company.first.update_attributes webhook: 'http://localhost:3031/webhooks/receive'
-  # - try wget << ipconfig getifaddr en0 >>:3031 from inside container
   def receive
-    # Step 1: save the incoming payload just in case further processing fails
-    @webhook = Webhook.create(payload: payload, origin: request.remote_ip)
```
Comment (amazing-jay) [inline app/controllers/webhooks_controller.rb:16]: return head :ok if webhok_topic is not scenario_session/completed

File: app/controllers/webhooks_controller.rb (hunk 12 above)
Comment (amazing-jay) [reply app/controllers/webhooks_controller.rb:8]: determine qa vs prod by the value of this value

File: app/controllers/webhooks_controller.rb [hunk 14]
```diff
@@ -1,29 +1,31 @@
...
+    webhook_topic = request.headers["X-Scope-Topic"]
+    if webhook_topic_params["node"]
+      user_id = webhook_topic_params["node"]["user"]["username"] if webhook_topic_params["node"]["user"]
+      email = webhook_topic_params["node"]["user"]["email"] if webhook_topic_params["node"]["user"]
+      object_id = webhook_topic_params["node"]["id"]
+      state = webhook_topic_params["node"]["state"]
+    end
+
+    @webhook = WebhookTopic.find_by(webhook_topic: webhook_topic, user_id: user_id)
+    if @webhook.present? || state != "FINISHED"
```
Comment (amazing-jay) [inline app/controllers/webhooks_controller.rb:12]: get rid of `|| state != 'FINISHED"` (so that this app doesn't break if the state is somehow wrong;

File: app/controllers/webhooks_controller.rb [hunk 15]
```diff
@@ -1,29 +1,31 @@
...
+      return head 500
```
Comment (amazing-jay) [inline app/controllers/webhooks_controller.rb:13]: `return head :ok`, (not 500) so that the server doesn't keep trying to send
